x-pack/plugins/lens/public/app_plugin/app.tsx

```diff
@@ -309,88 +318,144 @@ export function App({
-        getEsType: () => 'lens',
-        getDisplayName: () =>
-          i18n.translate('xpack.lens.app.saveModalType', {
-            defaultMessage: 'Lens visualization',
-          }),
-      },
-      saveProps.isTitleDuplicateConfirmed,
-      saveProps.onTitleDuplicate,
-      {
-        savedObjectsClient: core.savedObjects.client,
-        overlays: core.overlays,
-      }
-    );
-
-    const newlyCreated: boolean = saveProps.newCopyOnSave || !lastKnownDoc?.id;
-    docStorage
-      .save(doc)
-      .then(({ id }) => {
-        // Prevents unnecessary network request and disables save button
-        const newDoc = { ...doc, id };
-        const currentOriginatingApp = state.originatingApp;
-        setState((s) => ({
-          ...s,
-          isSaveModalVisible: false,
-          originatingApp:
-            newlyCreated && !saveProps.returnToOrigin ? undefined : currentOriginatingApp,
-          persistedDoc: newDoc,
-          lastKnownDoc: newDoc,
-        }));
-        if (docId !== id || saveProps.returnToOrigin) {
-          redirectTo(id, saveProps.returnToOrigin, newlyCreated);
+    const addedToLibrary = state.byValueMode && saveToLibrary;
+
+    const newlyCreated =
+      saveProps.newCopyOnSave || addedToLibrary || (!savedObjectId && !state.byValueMode);
+
+    if (state.byValueMode && !saveToLibrary && redirectToOrigin) {
+      await setState((s: LensAppState) => ({ ...s, persistedDoc: doc }));
+      const { savedObjectId: id, type, ...attributes } = doc;
+      redirectToOrigin({ attributes } as LensByValueInput);
+    } else {
+      await checkForDuplicateTitle(
+        {
+          ...doc,
+          copyOnSave: saveProps.newCopyOnSave,
+          lastSavedTitle: lastKnownDoc.title,
+          getEsType: () => 'lens',
+          getDisplayName: () =>
+            i18n.translate('xpack.lens.app.saveModalType', {
+              defaultMessage: 'Lens visualization',
+            }),
+        },
+        saveProps.isTitleDuplicateConfirmed,
+        saveProps.onTitleDuplicate,
+        {
+          savedObjectsClient: core.savedObjects.client,
+          overlays: core.overlays,
         }
-      })
-      .catch((e) => {
-        // eslint-disable-next-line no-console
-        console.dir(e);
-        trackUiEvent('save_failed');
-        core.notifications.toasts.addDanger(
-          i18n.translate('xpack.lens.app.docSavingError', {
-            defaultMessage: 'Error saving document',
-          })
-        );
-        setState((s) => ({ ...s, isSaveModalVisible: false }));
-      });
+      );
+      docStorage
+        .save(doc)
+        .then(({ savedObjectId: newSavedObjectId }) => {
+          // Prevents unnecessary network request and disables save button
+          const newDoc: Document = { ...doc, savedObjectId: newSavedObjectId };
+          const currentOriginatingApp = state.originatingApp;
+          setState((s) => ({
+            ...s,
+            byValueMode: false,
```

Comment:
It looks like you've added this line. I think this is handling the following case:

1. Edit a Lens visualization by value
2. Save a copy and don't return to the originating app
3. This is no longer editing by value

I wonder if this could be controlled by the URL, instead of being set here?

Reply:
The new version works this way - when a `savedObjectId` is specified in the url, it is not considered 'by value'. There is no by value mode in the state.